### ISSUE TITLE
No type checking on return type of arrow functions as parameter in a function.

### ISSUE THREAD

**Issue by MRNafisiA** [microsoft/TypeScript#47758]
# Bug Report

### 🔎 Search Terms
type-checking arrow-functions

### 🕗 Version & Regression Information

- No type checking is applied.
- I have tested it on 4.2.3, 4.3.5, 4.4.4, 4.5.4

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/C4TwDgpgBAChBOBnA9gOygXigbyqghgLYQBcUiw8AlqgOZQC+A3AFADGaFs1qwcSKVAElUAM2SYoACkhI0AIQCuVADYATBGSkBKTAD5YCQbowGABrMF4i0KoigASbJYXL1CHQDoCxBmdYsAPSBUACi8PDI8GQAKgDKAEwAzAkJseDQAOS4PqTklDS0THjIwAAiEKI0EGowkbKgZBQ8RYyZUHYlwFD4iIhUtAQARirQwBKgkFCZ-IKZ7JzdLqhks2iS2LlkmQBSyAAWqJkANF0VVag1dcgNINtU3Z2opT0qKsgA7jVQ+wjQ+Kg1B1HvYAPIAaU8mQYLCCIQAchJwpF4Cw6jQ+EZBCJxFIdPppJsbNs9ocTlAAI6KCAUKhobYffYgKBqZA0qAxDJxNjUMDdfBvT5QYD7TpVCDqAD80O02iYQA)

### 💻 Code

```ts
type Person = { name: string };
const PrintPerssonInfo = (personBuilder: () => Person) => `person name is ${personBuilder().name}`;

// Error: TS2322: Type '{ name: string; notDefinedProperty: string; }' is not assignable to type 'Person'
const person: Person = {name: 'John', notDefinedProperty: 'it is not allowed here and it is OK.'}

// No Error
PrintPerssonInfo(() => ({name: 'John', question: 'why does TypeScript allow this field?'}));
```

### 🙁 Actual behavior

I expected an error for defining 'question' key in Person type.

### 🙂 Expected behavior

calling function must have same error as defining the person.

**Comment by fatcerberus** [comment]
Excess property checks are not part of the type system proper and only apply to contextually-typed object literals.  In your case, you have an arrow function which is first inferred to be `() => { name: string, question: string }` (i.e. the return type is not `Person`, but instead is inferred directly from the expression) which is assignable to `() => Person` per structural subtyping.  Since the object literal is not directly assigned to a `Person`, the excess property check doesn't apply.

The error comes back if you write:
```ts
PrintPerssonInfo((): Person => ({name: 'John', question: 'this is an error again'}));
```

**Comment by MRNafisiA** [comment]
The problem stands for below code too
```
const w: () => Person = () => ({name: 'John', a: 'a'}); 
```
As you saying, We have to explicitly specify return type when ever we passing the function.
Is not it noisy? In my real case, return type is something big with multiple generic types. It is not a good idea to specify it explicitly on each call.
Is there an alternative approach or upcoming patch for this?

**Comment by fatcerberus** [comment]
Yes, it's the same issue.  The excess property check is only done directly between a fresh object literal and a known type, which isn't what's happening in your case: You have an arrow function without a return type annotation, so the compiler is free to infer the return type with no restriction; the type check that's ultimately done is between *two function types*--far too late for the compiler to do an excess property check.

Your cases are not errors for the same reason that
```ts
const getPersonWithExtraStuff = () => ({ name: "John", foo: "bar" });
const w: () => Person = getPersonWithExtraStuff;
```
is not an error.

This is by design.  Excess property checks exist largely as a linter check to catch some typos and are not meant to be relied on in general.  If you wanted to completely prevent excess properties at the type level, you would need #12936.

**Comment by MRNafisiA** [comment]
Thanks. It was very helpful.
